### PR TITLE
Java annotation array parsing fixes 

### PR DESCRIPTION
**Pull request by retronym**
- support trailing commas
  - support empty array initializer lists
  - when we hit syntax we don't support within an array initializer,
    brace match our way to the end of `}` that concludes the
    list.

### CODE REVIEW

File: test/files/neg/java-annotation-match-error/J.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+public class J {
+  @Ann(as = {
+    String.class,
+    class C,
+    String.class,
+    class D
+  })
+  public static void foo() {}
+
+  int int int
```
Comment (dwijnand) [inline test/files/neg/java-annotation-match-error/J.java]: Is this meant to be part of the test?

File: src/compiler/scala/tools/nsc/javac/JavaParsers.scala [hunk 2]
```diff
@@ -365,12 +365,15 @@ trait JavaParsers extends ast.parser.ParsersCommon with JavaScanners {
...
               atPos(in.pos) {
-                val elts = inBracesOrNil(commaSeparated(annVal()))
+                val elts = inBracesOrNil(if (in.token == RBRACE) Nil else commaSeparated(annVal()))
                 if (elts.exists(_.isEmpty)) EmptyTree
                 else Apply(ArrayModule_overloadedApply, elts: _*)
               }
             case _ if in.token == IDENTIFIER =>
               qualId(orClassLiteral = true)
+            case _ =>
+              syntaxError(in.currentPos, "Annotation value expected", skipIt = true)
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: I think following our reasoning on how non-constant values weren't supported before and aren't supported still, we should continue to just skip those, no?  No error?

Comment (lrytz) [reply src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: Yeah, the syntax allows a lot more https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-ElementValue

For example, this should not give a syntax error

```
public class A {
  public @interface Ann { int value(); } 
  @Ann((1 == 2) ? 3 : 4)
  public static void foo() { return; }
}
```

Comment (retronym) [reply src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: I'll amend and add more neg tests. The existing bailout mechanism doesn't deal with the unsupported syntax within array initializers.

Comment (retronym) [reply src/compiler/scala/tools/nsc/javac/JavaParsers.scala]: I also found an existing bug with annotation top level classes in the empty package:

```java
@Ann(values = {"foo", "bar"})
class Test {}

@interface Ann {
    String[] values();
}
```

```
sandbox/Test.java:2: error: `package' expected but `class' found.
class Test {}
^
sandbox/Test.java:6: error: identifier expected but eof found.
}
 ^
two errors found
```

Another edge case arises when trailing commas are uses in array initializers:

```java
package p1;

@Ann(values = {1, 2, })
class Test {}

@interface Ann {
    int[] values();
}
```

That's legal but gives a `MatchError` in 2.12.x (like the main test case in this PR). My current patch drops the arguments unnecessarily. I'll refine it to support the trailing comma. We're in good company though, `javac` also rejected the trailing commas [until 2012](https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6337964)!

**Review by dwijnand — APPROVED** [review]
Oh, wow.  Thanks for the deep surgery, the docs and the other test variants!
